### PR TITLE
feat: add support for CUDA

### LINKED ISSUE #104: Add CUDA support
Neptune should also support to run on CUDA.

### PR DESCRIPTION
**Pull request by vmx**
Neptune now also runs on CUDA. The OpenCL version is a bit slower than
previous versions, due to the removal of 64-bit limbs from the kernel.

Closed #104.

### CODE REVIEW

File: src/proteus/sources.rs [hunk 1]
```diff
@@ -42,24 +114,69 @@ fn poseidon_source(field: &str, derived_constants: DerivedConstants) -> String {
...
         v_rest_offset = v_rest_offset,
         sparse_matrixes_offset = sparse_matrixes_offset,
+        strength = strength,
     )
 }
 
-pub fn generate_program<Fr>(limb64: bool, derived_constants: DerivedConstants) -> String
+/// Returns the kernels source code for the given constants.
+///
+/// The constants can be genrated based on the the arity and the strength. The `derived_constants`
```
Comment (porcuquine) [inline src/proteus/sources.rs:123]: 'generated'

File: build.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
+/// The build script is needed to compile the CUDA kernel.
+
+#[cfg(feature = "cuda")]
+fn main() {
+    use std::path::PathBuf;
+    use std::process::Command;
+    use std::{env, fs};
+
+    use bellperson::bls::Fr;
```
Comment (dignifiedquire) [inline build.rs:9]: Just use `blstrs::Scalar` directly, the reexport will go away with the halo deps change anyway and then you don't need to depend on `bellperson` in the build.

File: src/proteus/gpu.rs [hunk 3]
```diff
@@ -187,37 +162,57 @@ where
...
-            .run()
-            .map_err(|e| Error::GpuError(format!("{:?}", e)))?;
-
-        let mut frs = vec![<Fr as Field>::zero(); num_hashes];
-        self.program
-            .read_into_buffer(&result_buffer, 0, &mut frs)
-            .map_err(|e| Error::GpuError(format!("{:?}", e)))?;
-        Ok(frs.to_vec())
+        let kernel_name = match (A::to_usize(), self.constants.strength()) {
+            #[cfg(feature = "arity2")]
```
Comment (dignifiedquire) [inline src/proteus/gpu.rs:166]: I wonder if a quick macro would make this nicer

File: src/proteus/sources.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/proteus/sources.rs:123]: leave it to @porcuquine to find that one typo :D

Comment (porcuquine) [reply src/proteus/sources.rs:123]: I realize it's a weak review on its own — but if I don't mention it when I see it, I probably won't find it again.

File: src/proteus/gpu.rs (hunk 3 above)
Comment (vmx) [reply src/proteus/gpu.rs:166]: I looked into it. But I didn't find a quick way to make it work nicely with the `#[cfg()]` attributes. So if there's a nice way that improves code clarity, I'll leave that to a subsequent PR.

**Review by dignifiedquire — APPROVED** [review]

**Comment by vmx** [comment]
I've run `gbench` on and RTX 3090 and the OpenCL performance on master and this PR is similar. This branch seems slightly slower, though it's within the noise range (some runs are faster than master branch runs). Running with CUDA is about 10% faster.

I think it's ready to be merged (I'd do a squash merge into a single commit).

**Pull request merged**